presto-router/src/main/java/com/facebook/presto/router/RouterPluginManager.java

```diff
@@ -0,0 +1,245 @@
+    private final File installedPluginsDir;
+    private final List<String> plugins;
+    private final ArtifactResolver resolver;
+    private final AtomicBoolean pluginsLoading = new AtomicBoolean();
+    private final AtomicBoolean pluginsLoaded = new AtomicBoolean();
+
+    @Inject
+    public RouterPluginManager(
+            PluginManagerConfig config,
+            PasswordAuthenticatorManager passwordAuthenticatorManager)
+    {
+        requireNonNull(config, "config is null");
+
+        this.installedPluginsDir = config.getInstalledPluginsDir();
+        if (config.getPlugins() == null) {
+            this.plugins = ImmutableList.of();
+        }
+        else {
+            this.plugins = ImmutableList.copyOf(config.getPlugins());
+        }
+        this.resolver = new ArtifactResolver(config.getMavenLocalRepository(), config.getMavenRemoteRepository());
+        this.passwordAuthenticatorManager = requireNonNull(passwordAuthenticatorManager, "passwordAuthenticatorManager is null");
+    }
+
+    public void loadPlugins()
+            throws Exception
+    {
+        if (!pluginsLoading.compareAndSet(false, true)) {
+            return;
+        }
+
+        for (File file : listFiles(installedPluginsDir)) {
+            if (file.isDirectory()) {
+                loadPlugin(file.getAbsolutePath());
+            }
+        }
+
+        for (String plugin : plugins) {
+            loadPlugin(plugin);
+        }
+
+        pluginsLoaded.set(true);
+    }
+
+    private void loadPlugin(String plugin)
+            throws Exception
+    {
+        log.info("-- Loading plugin %s --", plugin);
+        PluginClassLoader pluginClassLoader = buildClassLoader(plugin);
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(pluginClassLoader)) {
+            loadPlugin(pluginClassLoader);
+        }
+        log.info("-- Finished loading plugin %s --", plugin);
+    }
+
+    private void loadPlugin(PluginClassLoader pluginClassLoader)
+    {
+        ServiceLoader<Plugin> serviceLoader = ServiceLoader.load(Plugin.class, pluginClassLoader);
+        List<Plugin> plugins = ImmutableList.copyOf(serviceLoader);
+
+        if (plugins.isEmpty()) {
+            log.warn("No service providers of type %s", Plugin.class.getName());
+        }
+
+        for (Plugin plugin : plugins) {
+            log.info("Installing %s", plugin.getClass().getName());
+            installPlugin(plugin, pluginClassLoader::clone);
+        }
+    }
+
+    public void installPlugin(Plugin plugin, Supplier<ClassLoader> duplicatePluginClassLoaderFactory)
+    {
+        for (PasswordAuthenticatorFactory authenticatorFactory : plugin.getPasswordAuthenticatorFactories()) {
+            log.info("Registering password authenticator %s", authenticatorFactory.getName());
+            passwordAuthenticatorManager.addPasswordAuthenticatorFactory(authenticatorFactory);
+        }
+    }
```

Comment:
In addition to this, we should also load Presto Authenticators using PrestoAuthenticatorManager. Please refer PluginManager class